trin-core/src/portalnet/find/iterators/findcontent.rs

```diff
@@ -0,0 +1,715 @@
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+// This basis of this file has been taken from the rust-libp2p codebase:
+// https://github.com/libp2p/rust-libp2p
+
+use std::collections::btree_map::{BTreeMap, Entry};
+use std::time::Instant;
+
+use discv5::kbucket::{Distance, Key};
+
+use super::super::query_pool::QueryState;
+use super::query::{Query, QueryConfig, QueryPeer, QueryPeerState, QueryProgress};
+
+pub enum FindContentResponse<TNodeId> {
+    ClosestNodes(Vec<TNodeId>),
+    Content(Vec<u8>),
+}
+
+#[derive(Debug)]
+pub enum FindContentQueryResult<TNodeId> {
+    ClosestNodes(Vec<TNodeId>),
+    Content {
+        content: Vec<u8>,
+        closest_nodes: Vec<TNodeId>,
+    },
+}
+
+#[derive(Debug, Clone)]
+struct ContentAndPeer<TNodeId> {
+    content: Vec<u8>,
+    peer: TNodeId,
+}
+
+#[derive(Debug, Clone)]
+pub struct FindContentQuery<TNodeId> {
+    /// The target key we are looking for
+    target_key: Key<TNodeId>,
+
+    /// The instant when the query started.
+    started: Option<Instant>,
+
+    /// The current state of progress of the query.
+    progress: QueryProgress,
+
+    /// The closest peers to the target, ordered by increasing distance.
+    /// Assumes that no two peers are equidistant.
+    closest_peers: BTreeMap<Distance, QueryPeer<TNodeId>>,
+
+    /// The content possibly found by the query.
+    content: Option<ContentAndPeer<TNodeId>>,
+
+    /// The number of peers for which the query is currently waiting for results.
+    num_waiting: usize,
+
+    /// The configuration of the query.
+    config: QueryConfig,
+}
+
+impl<TNodeId> Query<TNodeId, FindContentResponse<TNodeId>, FindContentQueryResult<TNodeId>>
+    for FindContentQuery<TNodeId>
+where
+    TNodeId: Into<Key<TNodeId>> + Eq + Clone,
+{
+    fn target(&self) -> Key<TNodeId> {
+        self.target_key.clone()
+    }
+
+    fn started(&self) -> Option<Instant> {
+        self.started
+    }
+
+    fn start(&mut self, start: Instant) {
+        self.started = Some(start);
+    }
+
+    fn on_success(&mut self, peer: &TNodeId, peer_response: FindContentResponse<TNodeId>) {
+        if let QueryProgress::Finished = self.progress {
+            return;
+        }
+
+        let key: Key<TNodeId> = peer.clone().into();
+        let distance = key.distance(&self.target_key);
+
+        // Mark the peer's progress.
+        match self.closest_peers.entry(distance) {
+            Entry::Vacant(..) => return,
+            Entry::Occupied(mut entry) => match entry.get().state() {
+                QueryPeerState::Waiting(..) => {
+                    assert!(
```

Comment:
It looks like the 3 `assert!` statements in `FindContentQuery` are identical - seems like it could be pulled out into a shared method. But it also seems like a slightly more informative message could be useful. Just thinking through if we see a panic while running this live, messages that would help immediately differentiate between the 3 different fail cases could be nice to have.